deps/etcd-k8s.md

```diff
@@ -0,0 +1,314 @@
+# Service and Model Discovery Interface
+
+## Background
+
+This document is part of a series of docs attempting to replace the ETCD dependency in Dynamo using Kubernetes
+
+Source: https://github.com/ai-dynamo/enhancements/blob/neelays/runtime/XXXX-runtime-infrastructure.md
+
+ETCD usage in Dynamo can be categorized into the following:
+- Endpoint Instance Discovery (Current focus)
+- Model / Worker Discovery (Current focus)
+- Worker Port Conflict Resolution
+- Router State Sharing Synchronization
+
+## ServiceDiscovery Interface
+
+To de-couple Dynamo from ETCD, we define a minimal `ServiceDiscovery` interface that can be satisfied by different backends (etcd, kubernetes, etc). In Kubernetes environments, we will use the Kubernetes APIs to implement this interface.
+
+### Methods
+- `create_instance`(namespace: str, component: str, metadata: dict) -> Instance
+    - Creates an instance and persists associated immutable metadata. Does not mark the instance as ready.
+- `list_instances`(namespace: str, component: str) -> List[Instance]
+    - Lists all instances that match the given namespace and component.
+- `get_metadata`(namespace: str, component: str, instance_id: str) -> Instance
+    - Returns the metadata for an instance.
+- `subscribe`(namespace: str, component: str) -> EventStream
+    - Subscribes to events for the set of instances that match (namespace, component). Returns a stream of events giving signals for when instances that match the subscription are created, deleted, or readiness status changes.
+- `set_instance_status`(instance: Instance, status: str)
+    - Marks the instance as ready or not ready for traffic.
+- `get_instance_status`(instance: Instance) -> str
+    - Returns the status of the instance.
+<!-- Useful for MDC entry. Note tied to the instance lifecycle -->
+- `write_kv`(key: str, value: str)
+    - Writes data at a given key. Data not associated with any instance.
+- `read_kv`(key: str)
+    - Reads data at a given key. Data not associated with any instance.
```

Comment:
I'm not sure we want to provide generic key/value features. I think this should always be in the domain specific language of dynamo instances.